lte/gateway/c/connection_tracker/EventTracker.cpp

```diff
@@ -0,0 +1,255 @@
+  const struct nlattr** tb = (const struct nlattr**) data;
+  int type                 = mnl_attr_get_type(attr);
+
+  if (mnl_attr_type_valid(attr, CTA_MAX) < 0) return MNL_CB_OK;
+
+  switch (type) {
+    case CTA_TUPLE_ORIG:
+      if (mnl_attr_validate(attr, MNL_TYPE_NESTED) < 0) {
+        perror("mnl_attr_validate");
+        return MNL_CB_ERROR;
+      }
+      break;
+    case CTA_TIMEOUT:
+    case CTA_MARK:
+    case CTA_SECMARK:
+      if (mnl_attr_validate(attr, MNL_TYPE_U32) < 0) {
+        perror("mnl_attr_validate");
+        return MNL_CB_ERROR;
+      }
+      break;
+  }
+  tb[type] = attr;
+  return MNL_CB_OK;
+}
+
+static int data_cb(const struct nlmsghdr* nlh, void* data) {
+  struct nlattr* tb[CTA_MAX + 1] = {};
+  struct nfgenmsg* nfg = (struct nfgenmsg*) mnl_nlmsg_get_payload(nlh);
+  struct flow_information flow;
+  struct in_addr src_ip;
+  struct in_addr dst_ip;
+
+  //  if ((nlh->nlmsg_type & 0xFF) != IPCTNL_MSG_CT_DELETE) {
+  //    return 0;
+  //  }
+
+  mnl_attr_parse(nlh, sizeof(*nfg), data_attr_cb, tb);
+  if (tb[CTA_TUPLE_ORIG]) {
+    print_tuple(tb[CTA_TUPLE_ORIG], &flow);
+  }
+
+  src_ip.s_addr = flow.saddr;
+  dst_ip.s_addr = flow.daddr;
+
+  switch (nlh->nlmsg_type & 0xFF) {
+    case IPCTNL_MSG_CT_NEW:
+      if (nlh->nlmsg_flags & (NLM_F_CREATE | NLM_F_EXCL))
+        MLOG(MINFO) << "     [NEW] src=" << inet_ntoa(src_ip) << ":"
+                    << ntohs(flow.sport) << " dst=" << inet_ntoa(dst_ip) << ":"
+                    << ntohs(flow.dport) << " proto=" << flow.l4_proto;
+      else
+        printf("%9s ", "[UPDATE] \n");
+      break;
+    case IPCTNL_MSG_CT_DELETE:
+      MLOG(MINFO) << "[DESTROY] src=" << inet_ntoa(src_ip) << ":"
+                  << ntohs(flow.sport) << " dst=" << inet_ntoa(dst_ip) << ":"
+                  << ntohs(flow.dport) << " proto=" << flow.l4_proto;
+      break;
+  }
+  ((magma::PacketGenerator*) data)->send_packet(&flow);
+
+  return MNL_CB_OK;
+}
+
+namespace magma {
+
+EventTracker::EventTracker(std::shared_ptr<PacketGenerator> pkt_gen)
+    : pkt_gen_(pkt_gen) {}
+
+int EventTracker::init_conntrack_event_loop(void) {
+  struct mnl_socket* nl;
+  char buf[MNL_SOCKET_BUFFER_SIZE];
+  int ret;
+
+  nl = mnl_socket_open(NETLINK_NETFILTER);
+  if (nl == NULL) {
+    perror("mnl_socket_open");
+    exit(EXIT_FAILURE);
+  }
+
+  if (mnl_socket_bind(
+          nl,
+          NF_NETLINK_CONNTRACK_NEW |
+              // NF_NETLINK_CONNTRACK_UPDATE |
+              NF_NETLINK_CONNTRACK_DESTROY,
+          MNL_SOCKET_AUTOPID) < 0) {
+    perror("mnl_socket_bind");
+    exit(EXIT_FAILURE);
+  }
+
+  while (1) {
+    ret = mnl_socket_recvfrom(nl, buf, sizeof(buf));
+    if (ret == -1) {
+      perror("mnl_socket_recvfrom");
+      exit(EXIT_FAILURE);
+    }
+    ret = mnl_cb_run(buf, ret, 0, 0, data_cb, (void*) pkt_gen_.get());
+    if (ret == -1) {
+      perror("mnl_cb_run");
+      exit(EXIT_FAILURE);
+    }
+  }
+
+  mnl_socket_close(nl);
+
+  return 0;
+}
+
```

Comment:
This looks good.
Can you add basic UT to validate you are getting start and end event from conntractD.

Reply:
As this relies on conntrack to do stuff I think we'd need a integ test of sorts. I'll probably add it in a separate PR as I can't think of a simple one